rust/src/shards/gui/layouts/grid.rs

```diff
@@ -0,0 +1,318 @@
+    "UI.Grid"
+  }
+
+  fn help(&mut self) -> OptionalString {
+    OptionalString(shccstr!("Simple grid layout."))
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn inputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!(
+      "The value that will be passed to the Contents shards of the grid."
+    ))
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn outputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!("The output of this shard will be its input."))
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&GRID_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => self.contents.set_param(value),
+      1 => Ok(self.striped.set_param(value)),
+      2 => Ok(self.min_width.set_param(value)),
+      3 => Ok(self.max_width.set_param(value)),
+      4 => Ok(self.spacing.set_param(value)),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.contents.get_param(),
+      1 => self.striped.get_param(),
+      2 => self.min_width.get_param(),
+      3 => self.max_width.get_param(),
+      4 => self.spacing.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.requiring.clear();
+
+    // Add UI.Parents to the list of required variables
+    util::require_parents(&mut self.requiring, &self.parents);
+
+    Some(&self.requiring)
+  }
+
+  fn exposedVariables(&mut self) -> Option<&ExposedTypes> {
+    self.exposing.clear();
+
+    if util::expose_contents_variables(&mut self.exposing, &self.contents) {
+      Some(&self.exposing)
+    } else {
+      None
+    }
+  }
+
+  fn hasCompose() -> bool {
+    true
+  }
+
+  fn compose(&mut self, data: &InstanceData) -> Result<Type, &str> {
+    if !self.contents.is_empty() {
+      self.contents.compose(data)?;
+    }
+
+    // Always passthrough the input
+    Ok(data.inputType)
+  }
+
+  fn warmup(&mut self, ctx: &Context) -> Result<(), &str> {
+    self.parents.warmup(ctx);
+
+    if !self.contents.is_empty() {
+      self.contents.warmup(ctx)?;
+    }
+    self.striped.warmup(ctx);
+    self.min_width.warmup(ctx);
+    self.max_width.warmup(ctx);
+    self.spacing.warmup(ctx);
+
+    Ok(())
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    self.spacing.cleanup();
+    self.max_width.cleanup();
+    self.min_width.cleanup();
+    self.striped.cleanup();
+    if !self.contents.is_empty() {
+      self.contents.cleanup();
+    }
+
+    self.parents.cleanup();
+
+    Ok(())
+  }
+
+  fn activate(&mut self, context: &Context, input: &Var) -> Result<Var, &str> {
+    if self.contents.is_empty() {
+      return Ok(*input);
+    }
+
+    if let Some(ui) = util::get_current_parent(*self.parents.get())? {
+      let mut grid = egui::Grid::new(EguiId::new(self, 0));
```

Comment:
all those `new` in egui.. can't they by easily cached inside Shards memory? instead of wasting stack or hell even allocations potentially

Reply:
See comment below.